add type promotion support for sparse tensors

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30429 add type promotion support for sparse tensors**

also:
Fix a bug in uncoalesced division.

Fixes: https://github.com/pytorch/pytorch/issues/31449
by calling .contiguous() on indices in mul on cuda.

General approach here is that we:
* compute the common dtype based on input tensors
* error if the output tensor is specified and the common type can't be cast back to the output type (e.g. for inplace ops)
* convert input tensor (values) to the common dtype
* perform the op as normal (computing at the common dtype instead of the result type).
* convert/copy the result values back to that of the result tensor (for in-place ops).

For uncoalesced division we need to coalesce, because an integral tensor with values=[1,1] at the same index divided by 2 would give 1/2 + 1/2 =0 instead of 2/2=1.

Differential Revision: [D19143223](https://our.internmc.facebook.com/intern/diff/D19143223)

nairbv committed:
add type promotion support for sparse tensors

also fix a bug in uncoalesced division

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -153,14 +154,26 @@ SparseTensor pow_sparse_scalar(const SparseTensor& t, Scalar value) {
 }
 
+// inplace coalesce
+const SparseTensor& coalesce_(const SparseTensor& tensor) {
```
(carrying over from previous discussions) We should either:
- document that in-place division does a copy
- file an issue for it.

This is because, by definition, in-place operations save memory by not doing a copy. This is not explicitly stated in our documentation, but I think that is a falling in our documentation.

nairbv replied:
Filed an issue, I think it's out of scope for this PR since the documentation issue is also related to type promoting ops more general, not just for sparse.  E.g. if you do a float_tensor *= double_tensor it's going to create a temporary double result tensor, which it will then copy into the float tensor.

Ticket here:
https://github.com/pytorch/pytorch/issues/30533

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -172,18 +172,19 @@ def test_mixed_type_backward(self, device):
 
     def _get_test_tensor(self, device, dtype, remove_zeros=False):
-        shape = [20, 20, 20]
+        shape = [10, 10, 10]
```
(carrying on from previous discussion) We should try to reuse https://github.com/pytorch/pytorch/blob/0febff36acc4a906ca95a9d39382028b10f92c33/test/test_torch.py#L14132-L14168 as much as possible instead of running our own implementation. It is generally very useful to be able to get a random test tensor.

nairbv replied:
I'm still not entirely sold on this. Would you be OK refactoring to a common tensor generator as part of a separate PR? 

The _make_tensor version is far more simplistic, doesn't generate any negative values and few zeros, and has confusing handling of half tensors that I'd want to resolve before using outside test_torch. Fixing all that could mean changing a lot of existing tests unrelated to this PR.

zou3519 replied:
I think the best way to do it is to do the refactor in a PR and then have this diff depend on that PR.

> The _make_tensor version is far more simplistic, doesn't generate any negative values and few zeros, and has confusing handling of half tensors that I'd want to resolve before using outside test_torch. 

One alternative is to move your helper function to test_torch and have `make_tensor` call that.

I am proposing that we unify the idea of "getting a new random test tensor" and not duplicate the logic across the codebase. It sounds like to me your concern is about the mechanism in which we're combining the two functions; I haven't thought too much about what the right refactoring would look like.

zou3519 replied:
On second thought, you can do it as well in a follow-up PR. It would be good to file an issue about deduplicating the code and insert it as a comment in the code inline as a reminder.

nairbv replied:
https://github.com/pytorch/pytorch/issues/30611

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -430,6 +431,104 @@ def test_transpose(self, device):
         a = torch.tensor([[True, True], [False, True]], device=device)
         self.assertEqual(a.t() == 0, a.t() == False)  # noqa: E712
 
+    def _test_sparse_op_input_tensors(self, device, dtype, coalesced, zeros=True):
```
(carrying over from previous discussion) if we want this testing to be comprehensive, then we should test tensors of varying sparsities. Right now every tensor that is being tested is 0% sparse.

zou3519 replied:
In addition, it would be nice to get a single helper, `make_sparse_tensor`, so that it can be reused beyond binary operations

nairbv replied:
> every tensor that is being tested is 0% sparse

I agree more sparsity could help, but it's not 0%. The random numbers should include some 0's for most of these, and 0's are excluded from the sparse representation.

zou3519 replied:
Sorry, I meant that all of the dense (half, float, double) tensors are at 0% sparsity. It is improbable for a random dense tensor to have a value that is 0

nairbv replied:
added a mask to ensure sparsity

zou3519 reviewed (changes requested):
Copy-paste of the unresolved points on the previous PRs. I am still reading through the type promotion code.

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -430,6 +431,104 @@ def test_transpose(self, device):
+        self.assertEqual(s.is_coalesced(), coalesced)
+        return (t, s)
+
+    def _get_precision(self, dtype, coalesced):
```
You might want to discuss testing in a world with type promotion with @mruberry. If the plan is for all (or most ops) to support type promotion, then the traditional @dtype and @prec decorators don't work because they assume that each operation is being run with one dtype.

mruberry replied:
That's true, the @dtype and @prec decorators are a poor fit for type promotion tests. Oh well.

Guarding sparse tests behind onlyOnCPUAndCUDA is a good idea, though.

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -430,6 +431,104 @@ def test_transpose(self, device):
+            self.assertRaises(RuntimeError, lambda: op(s1, d2.view(d2.numel())[0].item()))
+
+    @onlyOnCPUAndCUDA
+    def test_sparse_ops(self, device):
```
How long does this test take?

nairbv replied:
almost a minute

zou3519 replied:
It's generally bad for a single unit test to take one minute. I know this single test is a conglomeration of some number of tests though (108? 400? not sure) but the longer the test is, the greater the risk for (1) flakiness (2) being deemed a "slow test" and being delegated to not run on PRs.

nairbv replied:
Split this into four calling functions, one each for add/sub/mul/div so that no one test runs so long.

zou3519 replied:
Are we doing more testing than necessary? Is there any way to cut down on the total test time?

Let's assume that we have an op `op`. There are maybe `N` tests for op, where N is the number of dtypes. If we implement type promotion on `op`, then we end up testing `N x N` cases. This means that in the future, if all of our operations support type promotion, then we will end up with a quadratic number of tests that we have today. This doesn't sound scalable to me.

zou3519 replied:
Do you have insight into how numpy does their testing for this?

zou3519 replied:
One thing I am thinking is that we could test all "dtype1-dtype2" combinations where "dtype1 = dtype2", test one type promotion combination (like float - double?) and run the rest in the slow ci.

nairbv replied:
It's a lot, but I think this actually is significant meaningful coverage, and I'm planning to make another pass later where I delete a number of tests that overlap with the added coverage here. We've missed a lot of issues (two that were caught by this test: atomic add and uncoalesced div) that only affected a single dtype, or only when uncoalesced, etc. 

We don't necessarily need to add every promoting op -- I'd think we want better coverage for ops as basic as arithmetic.

There might also not be so many binary ops, but if we find this gets too slow we could turn some of this off later too.

zou3519 replied:
I'm used to unit tests taking less than 1 second to run.

Because of that, I'm having a hard time evaluating whether or not adding 60s to testing is cause for worry or not (with each test taking 15s). How long does a typical run of test/test_torch.py take to run, in comparison?

nairbv replied:
the full `python test_torch.py` takes quite a while to run, not sure exactly but certainly more than a few minutes.

I can certainly understand the concern of adding a lot of time to the test runs. I weigh that partly against the amount of coverage gained though, and to me this seems like pretty substantial meaningful coverage.

zou3519 replied:
Do we need inputs of size 10x10x10 (1000 values) to run for all of the tests though? Could we get away with something like 5 x 5 x 5 (125 values)? I'm not actually sure how many random tensors we are generating

nairbv replied:
I agree 5x5x5 is probably also plenty.

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

[ghstack-poisoned]

zou3519 commented:
Quick ask: the code would be easier to review if the PR body had a summary of the techniques used in this PR. In particular, it would be helpful to know how each operator is implemented (for example, "add" is implemented with add_out_sparse_contiguous, add_out_sparse_noncontiguous, add_out_dense_sparse_{cpu|cuda}. When reviewing this I'm manually building up that list on a sheet of paper.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -327,27 +332,72 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
+Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTensor& sparse_, Scalar value);
+
+SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value) {
+  if (!t.is_sparse()) {
```
Why is `t` a SparseTensor if it's not sparse?

nairbv replied:
That's existing code, only moved not changed. It works because `using SparseTensor = Tensor;`

zou3519 replied:
Do you know why we do that? It might be nicer to change the declaration to use `Tensor` instead, if `SparseTensor` is used just for style purposes.

nairbv replied:
I have no idea, but if changing it I'd prefer to do that in a separate PR, since unrelated.

zou3519 replied:
Since we're modifying this anyways, it's a quick fix especially if it is correct. But feel free to file an issue and follow up on it.

nairbv replied:
> It might be nicer to change the declaration to use Tensor instead, if SparseTensor is used just for style purposes.

I do assume it's for style/annotation purposes. I think changing it to `Tensor` could be incorrect/confusing though. This function only handles add(sparse, sparse), and if you give it a `(dense, sparse)`, it redirects to a function that has the `(dense, sparse)` signature. It's not necessarily a stylistic choice I would have made, but I also don't think it's wrong as-is.

nairbv replied:
Right, that's all I'm saying is it's an orthogonal style choice across the whole file. I don't think we should change this one location -- if we're going to get rid of it we should remove all uses. I know it's confusing to see the ability to pass a dense tensor to a function annotated to handle sparse, but if we simply change it to `Tensor` in this one place then the annotations become inconsistent across the file and someone might think this function handles dense.

It may also be possible to get rid of these conditions/checks (or replace torch checks with internal asserts) since there have been more recent changes to dispatch (in native_functions we now have lines like `SparseCPU: add_sparse_` so this shouldn't be called incorrectly), but it'd take a little investigation to decide if this annotation is desirable and supportable in a more readable way, so I think we should consider it separately.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -170,8 +183,17 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
   AT_ASSERT(t.is_sparse());
 
   if (is_same_tensor(r, t)) {
+    // Can't divide an uncoalesced integral tensor accurately. e.g. for a sparse int tensor with value 6
```
Does mul_out_sparse_zerodim have the same problem?

zou3519 replied:
Nope, it doesn't because the math works out. Please ignore.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -170,8 +183,17 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
     r._values().div_(value);
   } else {
+    if (!t.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
+      coalesce_(t);
```
t is marked as a `const SparseTensor&`, which means that we shouldn't be modifying it in-place.

nairbv replied:
is this a general C++ convention / best practice for readability, or does modification of attributes of a const reference cause potential issues?

I was thinking of `const` as being info for the compiler, not as being an annotation for readers, and assumed that since this modification was allowed that it's fine. It's confusing to me because in JVM languages it's considered good practice to make a value "final" (or "val") even if the object it refers to is mutable.

Since the semantics of the data held in `t` don't change, doing the coalesce out of place seems wasteful (the caller may eventually need to coalesce `t` again), but I don't think I have the option of changing t to be non-const.

zou3519 replied:
Const correctness is using the keyword const to prevent `const` objects from being mutated. Const references should not be modified; when they do get modified the compiler should complain about it. Here is an overview of const correctness: https://isocpp.org/wiki/faq/const-correctness#overview-const .

It's both a best practice for readability and an annotation for the compiler: as a reader, I expect a const reference to not get modified and I expect the compiler to enforce it. Any deviation from those two is fishy. In this case, we've deviated from both of those points: we're modifying a const reference, AND the compiler isn't enforcing it!

Now, the reason why the compiler doesn't complain in this case is that `tensor._values()` returns a Tensor and one can arbitrarily mutate a Tensor. We are not implementing this correctly (but I haven't thought about what exactly needs to be done here).

Fixing `_values()` such that we catch cases like this is beyond the scope of this PR, but you're welcome to do it. Fixing the code to make it not look like a red flag to readers is important and should be done in this PR.

zou3519 replied:
> I was thinking of `const` as being info for the compiler, not as being an annotation for readers, and assumed that since this modification was allowed that it's fine. It's confusing to me because in JVM languages it's considered good practice to make a value "final" (or "val") even if the object it refers to is mutable.

I'm not familiar with the `final` keyword so I can't make a comparison here

> Since the semantics of the data held in `t` don't change, doing the coalesce out of place seems wasteful (the caller may eventually need to coalesce `t` again), but I don't think I have the option of changing t to be non-const.

Let's think about it from another perspective. Let's say the user is doing a division of an (integer) sparse tensor with an integer scalar. After this division, their original tensor got modified! 

There's an argument to made that coalescing a tensor in-place is not actually a modification, especially if we're trying to hide coalesced-ness from the user.

nairbv replied:
> I'm not familiar with the final keyword so I can't make a comparison here

From a language perspective, C++ const and java final variables are basically the same afaik. In java, `final` variables just can't be re-assigned (the meaning of final java methods vs const c++ functions is different though). Similar language semantics doesn't mean coding conventions / best practices are the same though, and I could conceive of C++ compilers making optimizations that rely on attributes of const objects not having been modified (if the language standard says that's allowed?).

It seems like if the language wanted to allow optimizations based on this, they wouldn't allow one to call non-const functions on non-mutable attributes of a const reference though, which they do allow (since we can call functions like `copy_` on `indices_`/`values_`).

> as a reader, I expect a const reference to not get modified and I expect the compiler to enforce it.

I think this might be a difference in convention. When you say "it" though we're referring specifically to attributes of the object referred to, not to the const reference itself.

> Now, the reason why the compiler doesn't complain...

ahh, interesting. I hadn't paid much attention to the constness of tensor functions. Makes sense.

> We are not implementing this correctly ...

These are considered "unsafe"/raw access so could be intentional? See the "note" on `uncoalesced` here:
https://pytorch.org/docs/stable/sparse.html

Is there a way to restrict calls to non-const functions on attributes of a const class?

> There's an argument to made that coalescing a tensor in-place is not actually a modification, especially if we're trying to hide coalesced-ness from the user.

Right, and I'm also questioning that, a use case described here:

https://isocpp.org/wiki/faq/const-correctness#mutable-data-members

The coalesce benefits the caller, and doesn't change the values of the tensor. It's sort of like caching -- I would expect that an object containing a cache could update its cache even if referred to by a const reference. In a sense a coalesce is a cache of what will eventually need to be materialized. We do expose `is_coalesced()`, but that seems like an implementation detail similar to if someone exposed an `is_cached()` on an object that might cache the result of internal computations on its attributes. 

From the note referenced above, the user isn't supposed to care about the internal representation and just consider it an optimization detail.

In terms of consistency though, I do see that functions like `set_coalesced` is not marked as const. We could have hypothetically chose to have `coalesced` be marked as `mutable` in SparseTensor and didn't, so that makes me more partial to the argument that we should take the performance hit and do this out of place here.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -170,8 +183,17 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
+    // Can't divide an uncoalesced integral tensor accurately. e.g. for a sparse int tensor with value 6
+    // represented as values=[3, 3], integer division by 2 would give values=[1, 1] => 2 instead
+    // of 6 / 2 => 3
+    if (!r.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
```
use a comment for raw boolean arguments: `isIntegralType(..., /*includeBool=*/True)`

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -153,14 +154,26 @@ SparseTensor pow_sparse_scalar(const SparseTensor& t, Scalar value) {
 }
 
+// inplace coalesce
+const SparseTensor& coalesce_(const SparseTensor& tensor) {
```
The declaration of this should be 
`SparseTensor& coalesce_(SparseTensor& tensor)` because we're modifying the tensor in-place.

zou3519 reviewed (changes requested):
(only looked at division code) Modifying a const reference seems fishy

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -458,17 +525,22 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
+  TORCH_CHECK(canCast(commonDtype, r.scalar_type()), "Can't convert result type ", commonDtype, " to output ", r.scalar_type());
+
+  Tensor t_values = t._values().to(commonDtype);
+  Tensor s_values = src._values().to(commonDtype);
```
Let's say we promoted the tensors instead of promoting the `values()` of the tensors here. Then we wouldn't have had to change anything else in the implementation. What are the benefits of promoting the `values()` instead of the tensors?

nairbv replied:
it's just simpler in aggregate, a small inplace change to existing code. existing code is:
```
Tensor t_values = t._values();
Tensor s_values = src._values();
```
This is:
```
Tensor t_values = t._values().to(commonDtype);
Tensor s_values = src._values().to(commonDtype);
```
The alternate change without trying to refactor anything else would be:
```
Tensor t_converted = t.to(commonDtype);
Tensor s_converted = src.to(commonDtype);
Tensor t_values = t_converted._values();
Tensor s_values = s_converted._values();
```
We still need to change the below because the op still needs to be done on `commonDtype` not the result type.  `r_values.scalar_type()` might be narrower than the common type, e.g. in `float_tensor *= double_tensor` the commonDtype is double but r_values.scalar_type() is float.

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

[ghstack-poisoned]

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -430,6 +431,123 @@ def test_transpose(self, device):
+            return 1e-3
+        return 1e-5
+
+    def _test_sparse_op(self, op_name, inplace, dtype1, dtype2, device, coalesced):
```
I still think the special casing in this function makes it hard to read. We should optimize for readability of the tests so that future readers don't have to go through all of the lines here to figure out what combination of (op, input1_sparsity, input2_sparsity) is supported.

If we had something like a table that showed exactly what we support:
```
tests = {
    ('add', [SPARSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('sub', [SPARSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('mul', [SPARSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('div', [SPARSE, SPARSE], UNSUPPORTED, ALL_DTYPES)

    ('add', [DENSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('sub', [DENSE, SPARSE], NO_ERROR, ALL_DTYPES)
    ('mul', [DENSE, SPARSE], UNSUPPORTED, ALL_DTYPES)
    ('div', [DENSE, SPARSE], UNSUPPORTED, ALL_DTYPES)

    ('add', [SPARSE, DENSE], UNSUPPORTED, ALL_DTYPES)
    ('sub', [SPARSE, DENSE], UNSUPPORTED, ALL_DTYPES)
    ('mul', [SPARSE, DENSE], UNSUPPORTED, ALL_DTYPES)
    ('div', [SPARSE, DENSE_NOZEROS], UNSUPPORTED, ALL_DTYPES)

    ('add', [SPARSE, SCALAR], UNSUPPORTED, ALL_DTYPES)
    ('sub', [SPARSE, SCALAR], UNSUPPORTED, ALL_DTYPES)
    ('mul', [SPARSE, SCALAR], NO_ERROR, EXCEPT_HALF)
    ('div', [SPARSE, SCALAR_NOZERO], NO_ERROR, EXCEPT_HALF)
}
```
then we could run each of those through a function `run_test(op, inputs, expected_behavior, dtypes)` that doesn't have to do special casing.

In order to construct this table, I spent 10 minutes reading through ~70 LOC in this test to understand what is supported or not. From this table, a future code reader will only have to read ~16 lines and understand in less than one minute what is supported. (If you asked me "Do we support sparse-dense multiplication", it's easy to figure that out with the table).

I proposed this before in https://github.com/pytorch/pytorch/pull/30317#discussion_r350339659 and you had some concerns about this approach that I'd like to address here. 

>  My main concern with this approach though would be having the list of tests hard-coded, since it leaves a lot of room for missing a case. 

I agree it is more difficult to write tests correctly in this style but we should be optimizing for readability. We write code once, but future contributors will read the code more than once. Code review can also help against missing cases.

> There's a fair bit that would also still need to be added too (e.g. can_cast errors in the next diff are dependent on the dtypes of the operands), so I think this would end up being just as much code. 

The main priority is readability, not LOC.

> Plus by being specific the comments can describe each particular case in the right place

To me, the table looks self documenting. If we want to put extra comments (I noticed that your code even has error messages, which is great!) we could just add them as comments to the table or even add the error messages into the table.

> and by re-using test tensors created for out of place operations, we save a bit of run-time over creating new test tensors for each test.

We can still generate inputs once in this approach. For example, `SPARSE` can just be a random sparse tensor that is generated once rather than an enum / string. 

What do you think? I'm open to more discussion and feedback. I'd really like to see less special-casing like this in our testing codebase.

nairbv replied:
> I agree it is more difficult to write tests correctly in this style

A big part of how I think about readability is the ease of maintaining correctness in the future, not just about the ability to consume the test-coverage passively (though being able to confirm full-coverage at a glance also has value). If I were to write the test in the suggested style, I'd want to add a test to test-the-test, i.e. to confirm that the static test config contained all possible combinations. When I think what that would look like, a test validating that a static config contains what it says it contains, it doesn't look very nice to me.

I could imagine instead structuring the table as a map where we look up the metadata about how/if a particular combination should be tested, and missing-metadata fails the test. The test would still get called in a loop over all possible options, enforcing that no combination is missing.

Even with a test-of-the-test or a dictionary test-metadata-lookup structure, this would introduce some gaps. E.g. using what I assume is a list of dtypes for `EXCEPT_HALF`, we'd presumably be dropping coverage for half on cuda. If instead it's an enum, the special casing code doesn't go away.

> The main priority is readability, not LOC.

Sure, but you seem to be saying that special-cases impairs readability. I'm saying the special casing code doesn't go away. `if inplace and not torch.can_cast(common_dtype, dtype1):` doesn't disappear just by abstracting away the list of supported ops. Some special cases are inherent to fully covering all possibilities.

Even the main thing that would be abstracted away, whether an op is supported, seems like it could be clearer to me as `# op(sparse,sparse)\n if op_name != 'div':`, specifying the exception vs scanning through a full list of possible combinations. The table approach could also become harder to read if we do end up deciding we want to add mixed-type sparse coverage to other ops and include them in this test. The test as-written could also get a lot simpler if we decide to implement all/most of the unsupported cases.

> then we could run each of those through a function run_test(op, inputs, expected_behavior, dtypes) that doesn't have to do special casing.

It abstracts some of the special casing, but I don't think it removes all of it.

> SPARSE can just be a random sparse tensor that is generated once rather than an enum / string.

both sparse and dense tensors still needs to be re-generated/cloned for in-place tests. We could add special handling in the tests to clone only for in-place, but these are the kinds of conditions that make re-writing this not-necessarily as simple/clear as you're hoping.

If the tests and code are correct and properly cover all possible cases, I'd argue this is more of a subjective stylistic difference we could iterate on and think about, not necessarily something so obfuscated that it should block the PR (though I'll probably still change the in-place coalesce on `t` for this PR).

I think the style you're favoring may be partly informed by what has worked reasonably well for `method_tests()` in common_method_invocations, though I'm not entirely sure that approach is ideal even there. If we do land, maintain, and iterate on this test a bit, we might get to an approach we prefer or that changes our thinking on other test coverage. We can always change this to the tabular common-method-invocations style later if it's really not working.

zou3519 replied:
> A big part of how I think about readability is the ease of maintaining correctness in the future, not just about the ability to consume the test-coverage passively (though being able to confirm full-coverage at a glance also has value). If I were to write the test in the suggested style, I'd want to add a test to test-the-test, i.e. to confirm that the static test config contained all possible combinations. 

We don't need a test to validate the static config. How do you see it changing in the future? There are only this many cases so we won't add things to it, only modify it or simplify it as we improve unsupported operations. Because of that a test to validate the config is overkill.

> Even with a test-of-the-test or a dictionary test-metadata-lookup structure, this would introduce some gaps. E.g. using what I assume is a list of dtypes for EXCEPT_HALF, we'd presumably be dropping coverage for half on cuda. If instead it's an enum, the special casing code doesn't go away.

Sure, we could add another column for 'CUDA', and double the size of the table. I missed that we're testing different dtypes for CUDA, which is another argument against the readability of the current tests. There's always more information that we can add to the table to avoid special casing. 

> Sure, but you seem to be saying that special-cases impairs readability. I'm saying the special casing code doesn't go away. if inplace and not torch.can_cast(common_dtype, dtype1): doesn't disappear just by abstracting away the list of supported ops. Some special cases are inherent to fully covering all possibilities.

I think you're misunderstanding my definition of "special casing". What I would like to remove is different behavior in the test for each operator.  I wouldn't consider ` if inplace and not torch.can_cast(common_dtype, dtype1): ` to be special casing: that is common to all of the tests, no matter the operator. 

Examples of things that aren't common to each operator: the type of input (division by zero is not allowed), what dypes are supported on which devices (this varies between operators).

>  The test as-written could also get a lot simpler if we decide to implement all/most of the unsupported cases.

You've pointed out the primary reason why there is special casing: to support testing for unsupported cases. Let me rephrase why I wanted the table approach: it makes it easier for future maintainers to see exactly what the unsupported cases are.

> If the tests and code are correct and properly cover all possible cases, I'd argue this is more of a subjective stylistic difference we could iterate on and think about, not necessarily something so obfuscated that it should block the PR (though I'll probably still change the in-place coalesce on t for this PR).

I'm fine with not blocking this PR on this point. However, I'm not done reading through the PR and I don't want to open too many threads at once because then I can't keep track of all of them.

> I think the style you're favoring may be partly informed by what has worked reasonably well for method_tests() in common_method_invocations, though I'm not entirely sure that approach is ideal even there.

We have a few places in the codebase where we use tables instead of having special logic to handle unsupported operations: 
- https://github.com/pytorch/pytorch/blob/master/test/cpp_api_parity/parity-tracker.md
- https://github.com/pytorch/pytorch/blob/4fd20c0816d00e445d96085b9dc2fa7a443159ff/test/test_torch.py#L11423-L11484

nairbv replied:
> How do you see it changing in the future?

All code is eventually modified. I could imagine us adding `pow`, comparison ops, `where`, any other binary op. I'm not sure what it will eventually become, this is just a start for mixed-type sparse support.

nairbv replied:
I'd still like to do it in a follow-up (or maybe stacked) PR, but I think we might both being happy if I add a table including only the failure cases. That way the op-specific conditions are (mostly) removed from the main test, but we retain the constraint on coverage and don't have a giant list of working cases that could nearly be generated by a for loop. Newly added ops would have coverage enforced. Also, as we implement the cases that currently fail, the table would shrink.

zou3519 replied:
I'm fine with having the tests in their current state. Feel free to send a followup if you'd like, I'm curious how the table including failure cases will look like

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

[ghstack-poisoned]

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -430,6 +431,123 @@ def test_transpose(self, device):
+        else:
+            s = t.to_sparse()
+            indices = torch.cat((s.indices(), s.indices()), 1)
+            values = torch.cat((s.values(), s.values()), 0)
```
nit: Might be easier to read if we used `s = s + s` (because sparse addition always appends the values and indices). Since we have the assertions below anyways to check that things are correct, if we ever change the implementation then we'll have an alert

nairbv replied:
I'd find that harder to read since it requires thinking about the implementation of add. cat is a bit more explicit about what's happening.

zou3519 replied:
The reason why I find `s = s + s` easier to read is because I don't have to reason about if this code is cat-ing the correct things together. A comment on top of it "the implementation of add is to just cat the values and indices" clears up that point, but this is just a nit.

zou3519 commented on aten/src/ATen/native/BinaryOps.h:
```diff
@@ -7,6 +7,23 @@ namespace at { struct TensorIterator; }
 
 namespace at { namespace native {
 
+inline void alpha_check(const ScalarType dtype, Scalar alpha) {
```
It's nice to know what the operation is in the error message. But since this is not making the existing code any worse than it is, this is not a blocker.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -153,14 +154,26 @@ SparseTensor pow_sparse_scalar(const SparseTensor& t, Scalar value) {
   return div_out_sparse_zerodim(self, self, value);
 }
 
+// inplace coalesce
```
nit: I don't think we need this comment. `coalesce_` means in-place coalesce already

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -153,14 +154,26 @@ SparseTensor pow_sparse_scalar(const SparseTensor& t, Scalar value) {
 }
 
+// inplace coalesce
+SparseTensor& coalesce_(SparseTensor& tensor) {
```
Should be `static` b/c this is a helper function that should not be called from other translation units.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -170,16 +183,26 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
-    r.resize_as_(t);
+    Tensor t_tmp = t;
+    if (!t.is_coalesced() && isIntegralType(r.scalar_type(), /*includeBool=*/true)) {
+      t_tmp = t.coalesce();
```
We can actually do this in-place, by first copying `t` to `r`, doing in-place coalesce on `r`, and finally performing the division. However, based on the implementation of `coalesce_` this doesn't save any extra time or memory (because our in-place coalesce actually does it out of place...) so I don't think this matters

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -207,7 +230,9 @@ Tensor add_sparse(const Tensor& self, const Tensor& other, Scalar alpha) {
               "add(sparse, dense) is not supported. Use add(dense, sparse) instead.");
-  Tensor result = at::empty({0}, self.options());
+  auto commonDtype = at::result_type(self, other);
+  alpha_check(commonDtype, alpha);
```
We now have new `alpha_check` and `sub_check` things added to sparse addition and sparse division, respectively. We should add a test (or two) that check that they work as expected (i.e., perform a sparse addition with an incorrect alpha and assert that the error message is what we expected).

nairbv replied:
These are used for dense tensors as well, so we do have test coverage through that code path

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -218,66 +243,47 @@ Tensor& add_sparse_(Tensor& self, const Tensor& other, Scalar alpha) {
-  AT_ASSERT(!t.is_cuda());  // the dispatch argument
-  TORCH_CHECK(!r.is_cuda(), "add: expected 'out' to be CPU tensor, but got CUDA tensor");
-  TORCH_CHECK(!src.is_cuda(), "add: expected 'other' to be a CPU tensor, but got a CUDA tensor");
+SparseTensor& add_out_sparse_contiguous(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value, ScalarType commonDtype) {
```
Is it a convention to call the promoted type the "commonDtype"? I feel like a better name is "promotedDtype" because the dtype isn't really "common" to the inputs; the inputs can be different dtypes.

nairbv replied:
We call it common_dtype in tensoriterator so I think this makes sense. In an earlier version of this PR I called it promoted_type, and there was some objection:

https://github.com/pytorch/pytorch/pull/26051#discussion_r327831374

zou3519 replied:
Sounds reasonable

zou3519 reviewed (commented):
some more comments

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -458,17 +526,22 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
   int64_t r_i = 0, t_i = 0, s_i = 0;
 
+  auto commonDtype = promoteTypes(t_.scalar_type(), src_.scalar_type());
+  TORCH_CHECK(canCast(commonDtype, r.scalar_type()), "Can't convert result type ", commonDtype, " to output ", r.scalar_type());
```
It's good to state the operator name here

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -458,17 +526,22 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
+  Tensor s_values = src._values().to(commonDtype);
+
+  std::vector<int64_t> size = t_values.sizes().vec();
+  size[0] = max_nnz;
```
Where is `size` used?

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -501,15 +574,15 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
     while (t_i < t_nnz && s_i < s_nnz) {
       if (!index_preamble()) continue;
-      r_values.select(0, r_i).addcmul_(t_values.select(0, t_i), s_values.select(0, s_i));
+      r_buffer.select(0, r_i).addcmul_(t_values.select(0, t_i), s_values.select(0, s_i));
```
Is type promotion implemented on `addcmul_`? If it is then we don't need to cast `t._values()` and `src._values()` to the promoted dtype.

nairbv replied:
not implemented

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -524,6 +597,9 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
   }
 
+  r.resize_as_(src);
+  Tensor r_values = r_buffer.to(r.scalar_type());
```
Could we just construct `r_values` to be the correct dtype and not do a `.to(...)` here?

nairbv replied:
I think you're looking at another one of those cases where for ops like `float_tensor *= double_tensor` we need r_buffer.scalar_type()==double to be converted to the result type r_values.scalar_type()==float. We perform ops at the common/higher type but in some cases allow downcasting to a narrower output type (according to the rules of torch.can_cast).

zou3519 reviewed (commented):
Here are some comments for sparse mul. Sparse sub and sparse div look fine, so what is left to review is sparse add.

nairbv commented on test/test_type_promotion.py:
```diff
@@ -430,6 +431,123 @@ def test_transpose(self, device):
+        if inplace and not torch.can_cast(common_dtype, dtype1):
+            self.assertRaises(RuntimeError, lambda: op(dense1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, sparse2))
+            self.assertRaises(RuntimeError, lambda: op(sparse1, dense2))
```
for (2), I think we'll need to refactor testing code around this more generally. There's precision logic elsewhere that doesn't work very well, often related to 'half' in particular, and a number of workarounds. e.g. `@precisionOverride` is sometimes used for handling half types, but is tied to the `@dtypes` annotation that wouldn't work for type promoting tests. If most cases are for half anyway, the default precision for half tests should just be lower at the source. Fixing it and consolidating the code everywhere seems outside the scope of this PR, and could be a large change, but I think that'll need to happen eventually.

(3) I'll try the smaller test tensors, which may help a bit. I think we can continue to iterate too though as we think of better approaches.

nairbv replied:
(1) is tricky since this is short-circuiting cases that are invalid regardless of the operator or combination of dense/sparse. For operations that aren't supported at all, an error will be thrown before reaching the casting error.

nairbv replied:
added a comment to try to clarify, not sure if you're happy enough with that for now. If I do another iteration with an explicit list of failure conditions, that might make this clearer too and easier to add the specific message without scrambling the rest of the test

mruberry replied:
You're right, @nairbv, we do need to get better precision handling (and that is outside the scope of this PR). Type promotion tests are particularly tricky since they inherently handle multiple dtypes.

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

[ghstack-poisoned]

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -327,27 +332,72 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
+}
+
+SparseTensor& add_out_sparse_non_contiguous(SparseTensor& r, const SparseTensor& t, const SparseTensor& src, Scalar value, ScalarType commonDtype) {
+    Tensor t_values = t._values().to(commonDtype);
```
Why do we need to cast to `commonDtype`? In the end we're not actually performing any addition, so it looks more efficient to:
1) Do the multiplication with the Scalar
2) cast both values to output dtype
3) cat the values together to produce the output values

I think the same suggestion can apply to the CUDA code.

nairbv replied:
I can see that there are a couple of cases (e.g. `half_tensor += double_tensor` or `float_tensor += double_tensor`) where this could be unnecessarily slower as-written. For such examples, your suggestion would be to convert the double_tensor to half instead of the half tensor to double before performing the cat. This would be equivalent (currently) because the final `to()` operation doesn't perform a coalesce, and so wouldn't have any loss of precision _vs_ the current implementation as-written here.

I think a better improvement would be to accept the performance hit but perform the addition at the commonDtype (as we do in dense ops) before converting back to the output type. Having this code be written in a way that explicitly has the operation being performed at the result dtype and not the common dtype seems inconsistent, and like it could cause confusion, even though it could be faster.

i.e., when we call `half_tensor += double_tensor`, the intermediate result at the commonDtype should get coalesced before being converted back to the lower precision result type (but preferably only when the destination dtype is in fact lower precision). This would be similar to what was changed for `div`, where the issue was more pronounced.

I think the best place to fix this though would be in the implementation of sparse `to` (via `copy_sparse_` in SparseTensor.cpp). I could imagine such a change being made eventually anyway and could be beneficial for operations like `float_tensor.copy_(double_tensor)`.

I think what we have here might be good-enough for now though -- I'd prefer to make a change to `to/copy_sparse_` in a follow-up PR in case that raises other issues or testing concerns, and because deciding when to coalesce could be tricky. Thoughts?

zou3519 replied:
I don't follow why `half_tensor += double_tensor` is more expensive.

The current code in this PR does:
- cast half_tensor -> double_tensor
- cat double_tensor with double_tensor
- cast double_tensor -> half_tensor.

With my suggestion:
- cast double_tensor -> half_tensor
- cat half_tensor with half_tensor

zou3519 replied:
Resolved offline: performance hit seems OK; also, when someone converts a sparse tensor to a lower-precision sparse tensor via .to(...), it might make sense to coalesce or warn that the new tensor might not be usable, but the decision to change those should be motivated by finding use cases for half_tensor += double_tensor.

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -327,27 +332,72 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
     // and we concat the indices and values tensors instead.
     AT_DISPATCH_ALL_TYPES(
-      s_values.scalar_type(), "add_out_sparse_cuda", [&] {
+      commonDtype, "add_out_sparse_cuda", [&] {
```
This is... CPU, right? Is the name here wrong?

nairbv replied:
yup, existing name is wrong, will fix it.

zou3519 reviewed (commented):
Read through the add implementation, had some questions

nairbv commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -395,26 +448,39 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
         dstBuffer = dstBuffer.select(0, indices_accessor[d][k]);
       }
-      Tensor srcBuffer = values.select(0, k);
+      Tensor srcBuffer = values_buffer.select(0, k);
```
that's something I find frustrating about C++ in general. I try to follow whatever I'm seeing in the code I'm modifying, but this function already contained indices_accessor and dstBuffer and it's hard to remember to check. Do we have any sort of general preference for style?

zou3519 replied:
It's not a C++ problem (the language doesn't have style guidelines), I think it is a pytorch problem. We seem to have a few different preferences for style:

- the pytorch api goes snake_case
- TensorIterator code as well as some of the JIT code I've read goes camelCase

In general I try to follow what the current style of the file is. There are no action items here; I'm just complaining about the (previously existing) inconsistency

nairbv replied:
camelCase does seem to be pretty common in C++ though, despite `std` mostly being snake_case. I think some languages are more opinionated about style, which keeps your own code and library code consistent.

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

General approach here is that we:
* compute the common dtype based on input tensors
* error if the output tensor is specified and the common type can't be cast back to the output type (e.g. for inplace ops)
* convert input tensor (values) to the common dtype
* perform the op as normal (computing at the common dtype instead of the result type).
* convert/copy the result values back to that of the result tensor (for in-place ops).

For uncoalesced division we need to coalesce, because an integral tensor with values=[1,1] at the same index divided by 2 would give 1/2 + 1/2 =0 instead of 2/2=1.


[ghstack-poisoned]

zou3519 reviewed (approved):
lgtm: Implementation looks fine (the sparse code itself doesn't read very well, but that's a problem with the sparse implementation), tests are comprehensive

zou3519 commented on test/test_type_promotion.py:
```diff
@@ -437,6 +438,126 @@ def test_transpose(self, device):
+                self._test_sparse_op(op_name, inplace, dtype1, dtype2, device, coalesced)
+
+    @onlyOnCPUAndCUDA
+    def test_sparse_add(self, device):
```
Out of curiosity, how long do these take to run now that we've changed the sizes to [5, 5, 5]?

nairbv replied:
I don't remember what the old results were but I don't think it made a big difference. TestTypePromotionCPU.test_sparse_add_cpu takes 6 seconds and the cuda variant takes 30. I think the cuda variant takes so long due to copying small tensors to CUDA, and we can probably improve on this by refactoring test tensor generation in https://github.com/pytorch/pytorch/issues/30611.

mruberry replied:
Thank you for marking these to only run on CPU and CUDA!

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

General approach here is that we:
* compute the common dtype based on input tensors
* error if the output tensor is specified and the common type can't be cast back to the output type (e.g. for inplace ops)
* convert input tensor (values) to the common dtype
* perform the op as normal (computing at the common dtype instead of the result type).
* convert/copy the result values back to that of the result tensor (for in-place ops).

For uncoalesced division we need to coalesce, because an integral tensor with values=[1,1] at the same index divided by 2 would give 1/2 + 1/2 =0 instead of 2/2=1.

Differential Revision: [D19143223](https://our.internmc.facebook.com/intern/diff/D19143223)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 19266c5:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=19266c565890742e23042a328aec8c82c216d11c).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3980677) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=75950910) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=75950910))
<details>
<summary>
<code>Dec 17 09:36:38 RuntimeError: test_autograd failed!</code>
</summary>

```
Dec 17 09:36:37 Ran 1024 tests in 283.739s 
Dec 17 09:36:37  
Dec 17 09:36:37 FAILED (errors=2, skipped=13, expected failures=1) 
Dec 17 09:36:37  
Dec 17 09:36:37 Generating XML reports... 
Dec 17 09:36:38 Traceback (most recent call last): 
Dec 17 09:36:38   File "test/run_test.py", line 456, in <module> 
Dec 17 09:36:38     main() 
Dec 17 09:36:38   File "test/run_test.py", line 449, in main 
Dec 17 09:36:38     raise RuntimeError(message) 
Dec 17 09:36:38 RuntimeError: test_autograd failed! 
Dec 17 09:36:38 + cleanup 
Dec 17 09:36:38 + retcode=1 
Dec 17 09:36:38 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

nairbv commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -395,26 +448,39 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
     auto indices_accessor = indices.accessor<int64_t, 2>();
     for (int64_t k = 0; k < sparse._nnz(); k++) {
-      Tensor dstBuffer = r;
+      Tensor& dstBuffer = resultBuffer;
```
@zou3519 This suggested change from Tensor to Tensor& wasn't in the original code, and seems to have caused unit tests to fail. I'm going to change this back.

zou3519 replied:
What’s wrong with it? We shouldn’t revert lines just because the unit tests fail; we should understand what the problem is.

nairbv replied:
This is just refraining from making an additional change to the underlying implementation of add. There are infinite possible changes we could make, this one didn't work, and if we did make it doesn't actually seem related to this PR either.

zou3519 replied:
It turns out that my original suggestion doesn't work because later on we're assigning a Tensor to dstBuffer:
```
for (int64_t d = 0; d < sparse.sparse_dim(); d++) {
    dstBuffer = dstBuffer.select(0, indices_accessor[d][k]);
}
```
Assigning the rvalue Tensor to a Tensor& is wrong because the Tensor gets cleaned up instantly, leading to the reference dstBuffer pointing to garbage. Changing it back here is correct.

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

General approach here is that we:
* compute the common dtype based on input tensors
* error if the output tensor is specified and the common type can't be cast back to the output type (e.g. for inplace ops)
* convert input tensor (values) to the common dtype
* perform the op as normal (computing at the common dtype instead of the result type).
* convert/copy the result values back to that of the result tensor (for in-place ops).

For uncoalesced division we need to coalesce, because an integral tensor with values=[1,1] at the same index divided by 2 would give 1/2 + 1/2 =0 instead of 2/2=1.

Differential Revision: [D19143223](https://our.internmc.facebook.com/intern/diff/D19143223)

[ghstack-poisoned]

nairbv committed:
Update on "add type promotion support for sparse tensors"


also fix a bug in uncoalesced division

General approach here is that we:
* compute the common dtype based on input tensors
* error if the output tensor is specified and the common type can't be cast back to the output type (e.g. for inplace ops)
* convert input tensor (values) to the common dtype
* perform the op as normal (computing at the common dtype instead of the result type).
* convert/copy the result values back to that of the result tensor (for in-place ops).

For uncoalesced division we need to coalesce, because an integral tensor with values=[1,1] at the same index divided by 2 would give 1/2 + 1/2 =0 instead of 2/2=1.

Differential Revision: [D19143223](https://our.internmc.facebook.com/intern/diff/D19143223)

[ghstack-poisoned]
